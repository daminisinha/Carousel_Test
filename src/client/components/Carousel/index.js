import React, { Component } from 'react';
import { Container, Row, Col, Hidden, Visible } from 'react-grid-system';
import Slide from '../Slide'
import Thumbnails from '../Thumbnails';
import NavigateButtonContainer from '../NavigateButtonContainer'
import RightArrow from '../RightArrow'
import LeftArrow from '../LeftArrow'
import './index.scss'


export default class Carousel extends Component {
    constructor(props) {
        super(props);
        this.state = { activeIndex: 0 };
        this.nextImage = this.nextImage.bind(this);
        this.prevImage = this.prevImage.bind(this);
        this.setCurrentImageIndex = this.setCurrentImageIndex.bind(this)
    }
    nextImage(e) {
        if (this.props.images.length - 1 === this.state.activeIndex) {
            this.setState({ activeIndex: 0 })
        } else {
            this.setState({ activeIndex: this.state.activeIndex + 1 })
        }
    }

    prevImage(e) {
        if (0 === this.state.activeIndex) {
            this.setState({ activeIndex: this.props.images.length - 1 })

        } else {
            this.setState({ activeIndex: this.state.activeIndex - 1 })
        }

    }
    setCurrentImageIndex(index) {

        if (index !== 'undefined' && index >= 0) {
            this.setState({ activeIndex: index })
        }

    }
    render() {
        let image = this.props.images[this.state.activeIndex]
        return (

            <Container fluid={true} className="carousel-container">
                <Thumbnails images={this.props.images} onClickHandler={this.setCurrentImageIndex} />
                <Row justify="center">
                    <Col xs={12} sm={12} lg={10} md={10}>

                        <ul className="carousel">

                            <li className="center-slide">
                            <Visible xs sm md>
                                    <LeftArrow prevImage={this.prevImage} />
                                </Visible>
                              <Slide id="contentCarousel"
                                    image={image}
                                    width="100%"
                                    height="480px"
                                    index={this.state.activeIndex}
                                    onClickHandler={this.setCurrentImageIndex}
                                />
                                <Visible xs sm md>
                                    <RightArrow nextImage={this.nextImage} />
                                </Visible>
                                <h3>By {image.user}</h3>
                            </li>
                        </ul>
                        <NavigateButtonContainer prevImage={this.prevImage} nextImage={this.nextImage} />
                    </Col>
                </Row>
            </Container>
        );
    }
}

