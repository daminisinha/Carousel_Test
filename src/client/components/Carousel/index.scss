@import '../../utilities/variables';
.carousel {
    display:flex;
    flex-direction: row;
    justify-content: center;
    list-style-type: none;
    white-space: nowrap;
    padding:0px;
    position:relative;
    text-align: center;
    margin-top:0px;
}

.carousel-container{
    background-color:white;
    width:100%;
    margin:0px !important;
}

.carousel>li{
    padding-bottom:20x;
    margin:10px; 
}

#contentCarousel{
    width:100%;
    max-width: 100%;
    max-height:80%;
    
}

.center-slide{
    flex-grow:2;
    display:block; 
    padding-top:0px;
}
