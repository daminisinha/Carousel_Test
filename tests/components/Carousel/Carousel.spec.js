import React from 'react';
import Carousel from '../../../src/client/components/Carousel'
import LeftArrow from '../../../src/client/components/LeftArrowButton';
import RightArrow from '../../../src/client/components/RightArrowButton';
import ImageNavButton from '../../../src/client/components/NavigateButton'
import {shallow} from 'enzyme';
import imagesObject from '../../setup/dummy';
const images = imagesObject.hits.slice(0,6);
describe('Carousel Tests',()=>{
    const component = shallow(<Carousel images={images}/>); 
    it('Carousel renders',()=>{
        
        expect(component.find('Container')).toHaveLength(1);
        expect(component.find('Thumbnails')).toHaveLength(1);
        expect(component.find('LeftArrow')).toHaveLength(0);
        expect(component.find('RightArrow')).toHaveLength(0);
        expect(component.find('Slide')).toHaveLength(0);
        expect(component.find('NavigateButtonContainer')).toHaveLength(1);
        expect(component).toMatchSnapshot();
    })
    it('Check left arrow click',()=>{
        const prevImage = jest.fn();
        const leftArrow = shallow(<LeftArrow prevImage={prevImage}/>)
        leftArrow.find('img').simulate('click');
        expect(prevImage.mock.calls.length).toBe(1)
    })
    it('Check right arrow click',()=>{
        const nextImage = jest.fn();
        const rightArrow = shallow(<RightArrow nextImage={nextImage}/>)
        rightArrow.find('img').simulate('click');
        expect(nextImage.mock.calls.length).toBe(1)
    })
    it('Check button clicks for changing image',()=>{
        const imageChangeHandler = jest.fn();
        const button = shallow(<ImageNavButton buttonText="Next" onClickHandler={imageChangeHandler}/>)
        button.simulate('click');
        expect(imageChangeHandler.mock.calls.length).toBe(1)
    })
})

