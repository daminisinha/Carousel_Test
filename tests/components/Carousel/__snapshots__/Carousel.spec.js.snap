// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel Tests Carousel renders 1`] = `
<Container
  className="carousel__container"
  component="div"
  fluid={true}
  lg={false}
  md={false}
  sm={false}
  style={Object {}}
  xl={false}
  xs={false}
>
  <Thumbnails
    images={
      Array [
        Object {
          "comments": 48,
          "downloads": 2866,
          "favorites": 14,
          "id": 3841703,
          "imageHeight": 3271,
          "imageSize": 3944579,
          "imageWidth": 4900,
          "largeImageURL": "https://pixabay.com/get/ea3db50e2ff4023ed1584d05fb1d4f90e671e2d31cac104491f3c77eaeefb0b0_1280.jpg",
          "likes": 50,
          "pageURL": "https://pixabay.com/en/leaf-water-surface-floating-3841703/",
          "previewHeight": 100,
          "previewURL": "https://cdn.pixabay.com/photo/2018/11/27/15/53/leaf-3841703_150.jpg",
          "previewWidth": 150,
          "tags": "leaf, water, surface",
          "type": "photo",
          "user": "MabelAmber",
          "userImageURL": "https://cdn.pixabay.com/user/2018/11/02/20-20-42-202_250x250.jpg",
          "user_id": 1377835,
          "views": 3354,
          "webformatHeight": 427,
          "webformatURL": "https://pixabay.com/get/ea3db50e2ff4023ed1584d05fb1d4f90e671e2d31cac104491f3c77eaeefb0b0_640.jpg",
          "webformatWidth": 640,
        },
        Object {
          "comments": 42,
          "downloads": 3326,
          "favorites": 16,
          "id": 3842408,
          "imageHeight": 3246,
          "imageSize": 2244944,
          "imageWidth": 3246,
          "largeImageURL": "https://pixabay.com/get/ea3db50d2cf4093ed1584d05fb1d4f90e671e2d31cac104491f3c77eaeefb0b0_1280.jpg",
          "likes": 48,
          "pageURL": "https://pixabay.com/en/cat-animal-mammal-feline-portrait-3842408/",
          "previewHeight": 150,
          "previewURL": "https://cdn.pixabay.com/photo/2018/11/27/21/21/cat-3842408_150.jpg",
          "previewWidth": 150,
          "tags": "cat, animal, mammal",
          "type": "photo",
          "user": "MabelAmber",
          "userImageURL": "https://cdn.pixabay.com/user/2018/11/02/20-20-42-202_250x250.jpg",
          "user_id": 1377835,
          "views": 3850,
          "webformatHeight": 640,
          "webformatURL": "https://pixabay.com/get/ea3db50d2cf4093ed1584d05fb1d4f90e671e2d31cac104491f3c77eaeefb0b0_640.jpg",
          "webformatWidth": 640,
        },
        Object {
          "comments": 49,
          "downloads": 9293,
          "favorites": 24,
          "id": 3834860,
          "imageHeight": 3264,
          "imageSize": 4104892,
          "imageWidth": 4896,
          "largeImageURL": "https://pixabay.com/get/ea3db20b20f2013ed1584d05fb1d4f90e671e2d31cac104491f3c77eaeefb0b0_1280.jpg",
          "likes": 78,
          "pageURL": "https://pixabay.com/en/christmas-motif-building-3834860/",
          "previewHeight": 99,
          "previewURL": "https://cdn.pixabay.com/photo/2018/11/24/01/35/christmas-motif-3834860_150.jpg",
          "previewWidth": 150,
          "tags": "christmas motif, building, fachwerkhÃ¤user",
          "type": "photo",
          "user": "Couleur",
          "userImageURL": "https://cdn.pixabay.com/user/2018/09/29/00-20-33-645_250x250.jpg",
          "user_id": 1195798,
          "views": 11839,
          "webformatHeight": 426,
          "webformatURL": "https://pixabay.com/get/ea3db20b20f2013ed1584d05fb1d4f90e671e2d31cac104491f3c77eaeefb0b0_640.jpg",
          "webformatWidth": 640,
        },
        Object {
          "comments": 35,
          "downloads": 1118,
          "favorites": 6,
          "id": 3816193,
          "imageHeight": 2671,
          "imageSize": 2660531,
          "imageWidth": 4000,
          "largeImageURL": "https://pixabay.com/get/ea3db00929fd023ed1584d05fb1d4f90e671e2d31cac104491f3c77eaeefb0b0_1280.jpg",
          "likes": 44,
          "pageURL": "https://pixabay.com/en/hedgehog-spur-hannah-foraging-3816193/",
          "previewHeight": 100,
          "previewURL": "https://cdn.pixabay.com/photo/2018/11/14/22/12/hedgehog-3816193_150.jpg",
          "previewWidth": 150,
          "tags": "hedgehog, spur, hannah",
          "type": "photo",
          "user": "Capri23auto",
          "userImageURL": "https://cdn.pixabay.com/user/2018/05/16/20-25-22-995_250x250.jpg",
          "user_id": 1767157,
          "views": 1289,
          "webformatHeight": 427,
          "webformatURL": "https://pixabay.com/get/ea3db00929fd023ed1584d05fb1d4f90e671e2d31cac104491f3c77eaeefb0b0_640.jpg",
          "webformatWidth": 640,
        },
        Object {
          "comments": 38,
          "downloads": 1760,
          "favorites": 13,
          "id": 3841692,
          "imageHeight": 3271,
          "imageSize": 4020542,
          "imageWidth": 4900,
          "largeImageURL": "https://pixabay.com/get/ea3db50e2efd033ed1584d05fb1d4f90e671e2d31cac104491f3c77eaeefb0b0_1280.jpg",
          "likes": 39,
          "pageURL": "https://pixabay.com/en/leaf-vein-pattern-twig-bush-3841692/",
          "previewHeight": 100,
          "previewURL": "https://cdn.pixabay.com/photo/2018/11/27/15/51/leaf-3841692_150.jpg",
          "previewWidth": 150,
          "tags": "leaf, vein, pattern",
          "type": "photo",
          "user": "MabelAmber",
          "userImageURL": "https://cdn.pixabay.com/user/2018/11/02/20-20-42-202_250x250.jpg",
          "user_id": 1377835,
          "views": 2038,
          "webformatHeight": 427,
          "webformatURL": "https://pixabay.com/get/ea3db50e2efd033ed1584d05fb1d4f90e671e2d31cac104491f3c77eaeefb0b0_640.jpg",
          "webformatWidth": 640,
        },
        Object {
          "comments": 34,
          "downloads": 1009,
          "favorites": 11,
          "id": 3840264,
          "imageHeight": 3387,
          "imageSize": 4846269,
          "imageWidth": 5353,
          "largeImageURL": "https://pixabay.com/get/ea3db50f2af2053ed1584d05fb1d4f90e671e2d31cac104491f3c77eaeefb0b0_1280.jpg",
          "likes": 43,
          "pageURL": "https://pixabay.com/en/christmas-angel-hope-3840264/",
          "previewHeight": 94,
          "previewURL": "https://cdn.pixabay.com/photo/2018/11/26/21/07/christmas-3840264_150.jpg",
          "previewWidth": 150,
          "tags": "christmas, angel, hope",
          "type": "photo",
          "user": "KleineKiwi",
          "userImageURL": "https://cdn.pixabay.com/user/2018/11/14/09-23-35-829_250x250.jpg",
          "user_id": 9883074,
          "views": 1392,
          "webformatHeight": 404,
          "webformatURL": "https://pixabay.com/get/ea3db50f2af2053ed1584d05fb1d4f90e671e2d31cac104491f3c77eaeefb0b0_640.jpg",
          "webformatWidth": 640,
        },
      ]
    }
    onClickHandler={[Function]}
  />
  <Row
    align="normal"
    component="div"
    debug={false}
    gutterWidth={null}
    justify="center"
    nogutter={false}
    nowrap={false}
    style={Object {}}
  >
    <Col
      component="div"
      debug={false}
      lg={10}
      md={10}
      offset={Object {}}
      pull={Object {}}
      push={Object {}}
      sm={12}
      style={Object {}}
      xl={null}
      xs={12}
    >
      <ul
        className="carousel__list_style"
      >
        <li
          className="list_style_slide"
        >
          <Visible
            lg={false}
            md={true}
            sm={true}
            xl={false}
            xs={true}
          >
            <LeftArrowButton
              prevImage={[Function]}
            />
          </Visible>
          <Slider
            height="480px"
            id="contentCarousel"
            image={
              Object {
                "comments": 48,
                "downloads": 2866,
                "favorites": 14,
                "id": 3841703,
                "imageHeight": 3271,
                "imageSize": 3944579,
                "imageWidth": 4900,
                "largeImageURL": "https://pixabay.com/get/ea3db50e2ff4023ed1584d05fb1d4f90e671e2d31cac104491f3c77eaeefb0b0_1280.jpg",
                "likes": 50,
                "pageURL": "https://pixabay.com/en/leaf-water-surface-floating-3841703/",
                "previewHeight": 100,
                "previewURL": "https://cdn.pixabay.com/photo/2018/11/27/15/53/leaf-3841703_150.jpg",
                "previewWidth": 150,
                "tags": "leaf, water, surface",
                "type": "photo",
                "user": "MabelAmber",
                "userImageURL": "https://cdn.pixabay.com/user/2018/11/02/20-20-42-202_250x250.jpg",
                "user_id": 1377835,
                "views": 3354,
                "webformatHeight": 427,
                "webformatURL": "https://pixabay.com/get/ea3db50e2ff4023ed1584d05fb1d4f90e671e2d31cac104491f3c77eaeefb0b0_640.jpg",
                "webformatWidth": 640,
              }
            }
            index={0}
            onClickHandler={[Function]}
            width="100%"
          />
          <Visible
            lg={false}
            md={true}
            sm={true}
            xl={false}
            xs={true}
          >
            <RightArrowButton
              nextImage={[Function]}
            />
          </Visible>
          <h3>
            Author: 
            MabelAmber
          </h3>
        </li>
      </ul>
      <NavigateButtonContainer
        nextImage={[Function]}
        prevImage={[Function]}
      />
    </Col>
  </Row>
</Container>
`;
